programs/wen_new_standard/src/instructions/mint/royalties/add.rs

```diff
@@ -91,14 +98,88 @@ pub fn handler(ctx: Context<AddRoyalties>, args: UpdateRoyaltiesArgs) -> Result<
+                let bump = verify_extra_meta_account(
+                    &mint.key(),
+                    &extra_metas_account.key(),
+                    &TRANSFER_GUARD_PROGRAM_ID,
+                )?;
+                (TRANSFER_GUARD_PROGRAM_ID, bump)
+            } else {
+                let bump = verify_extra_meta_account(
+                    &mint.key(),
+                    &extra_metas_account.key(),
+                    &WNS_PROGRAM_ID,
+                )?;
+                (WNS_PROGRAM_ID, bump)
+            }
+        }
+        None => {
+            let bump = verify_extra_meta_account(
+                &mint.key(),
+                &extra_metas_account.key(),
+                &WNS_PROGRAM_ID,
+            )?;
+            (WNS_PROGRAM_ID, bump)
+        }
+    };
+
+    // add metadata program/transfer guard program as the transfer hook program
+    ctx.accounts
+        .update_transfer_hook_program_id(transfer_hook_program_id)?;
+
+    if transfer_hook_program_id.eq(&WNS_PROGRAM_ID) {
+        let signer_seeds: &[&[&[u8]]] = &[&[META_LIST_ACCOUNT_SEED, mint_key.as_ref(), &[bump]]];
+        let account_size =
+            ExtraAccountMetaList::size_of(get_meta_list_size(get_approve_account_pda(mint_key)))?;
+        let lamports = Rent::get()?.minimum_balance(account_size);
+
+        // create ExtraAccountMetaList account
+        create_account(
+            CpiContext::new(
+                ctx.accounts.system_program.to_account_info(),
+                CreateAccount {
+                    from: ctx.accounts.payer.to_account_info(),
+                    to: ctx.accounts.extra_metas_account.to_account_info(),
+                },
+            )
+            .with_signer(signer_seeds),
+            lamports,
+            account_size as u64,
+            ctx.program_id,
+        )?;
+
+        // initialize the extra metas account
+        let metas = get_meta_list(get_approve_account_pda(mint.key()));
+        let mut data = extra_metas_account.try_borrow_mut_data()?;
+        ExtraAccountMetaList::init::<ExecuteInstruction>(&mut data, &metas)?;
+    } else {
+        let transfer_guard_program_info = transfer_hook_program_account_info.unwrap();
+        let guard_account_info = ctx.remaining_accounts.get(1);
+        if guard_account_info.is_none() {
+            return err!(ExtraAccountMetaListErrors::InvalidGuardAccount);
+        }
+        let guard_account_info = guard_account_info.unwrap();
+        initialize(CpiContext::new(
+            transfer_guard_program_info.clone(),
+            Initialize {
+                extra_metas_account: extra_metas_account.to_account_info(),
+                guard: guard_account_info.clone(),
+                mint: mint.to_account_info(),
+                payer: payer.to_account_info(),
+                system_program: system_program.to_account_info(),
+                transfer_hook_authority: authority.to_account_info(),
+            },
+        ))?;
+    }
```

Comment:
I'm wondering if we can break this up some so more is shared?

The both ineed to create an extrr_meta_account but is size is different between the 2.
It both needs a list of pubkeys to save to the extra metadata see if the condition is more around getting this list and less on the creation of the extra_meta_account.